gateway/src/apicast/template_string.lua

```diff
@@ -0,0 +1,87 @@
+local Liquid = require 'liquid'
+local LiquidTemplate = Liquid.Template
+local LiquidInterpreterContext = Liquid.InterpreterContext
+local LiquidFilterSet = Liquid.FilterSet
+local LiquidResourceLimit = Liquid.ResourceLimit
+local ngx = ngx
+
+local setmetatable = setmetatable
+local pairs = pairs
+
+local LiquidTemplateString = {}
+local liquid_template_string_mt = { __index = LiquidTemplateString }
+
+-- Expose only ngx.* functions that we think could be useful and that do not
+-- have any side-effects.
+local liquid_filters = {
+  ngx_excape_uri = ngx.escape_uri,
+  ngx_unescape_uri = ngx.unescape_uri,
+  ngx_encode_base64 = ngx.encode_base64,
+  ngx_decode_base64 = ngx.decode_base64,
+  crc32_short = ngx.crc32_short,
+  ngx_crc32_long = ngx.crc32_long,
+  ngx_hmac_sha1 = ngx.hmac_sha1,
+  ngx_md5 = ngx.md5,
+  ngx_md5_bin = ngx.md5_bin,
+  ngx_sha1_bin = ngx.sha1_bin,
+  ngx_quote_sql_str = ngx.quote_sql_str,
+  ngx_today = ngx.today,
+  ngx_time = ngx.time,
+  ngx_now = ngx.now,
+  ngx_localtime = ngx.localtime,
+  ngx_utctime = ngx.utctime,
+  ngx_cookie_time = ngx.cookie_time,
+  ngx_http_time = ngx.http_time,
+  ngx_parse_http_time = ngx.parse_http_time
+}
+
+local liquid_filter_set = LiquidFilterSet:new()
+for name, func in pairs(liquid_filters) do
+  liquid_filter_set:add_filter(name, func)
+end
+
+-- Set resource limits to avoid loops
+local liquid_resource_limit = LiquidResourceLimit:new(nil, nil, 0)
+
+function LiquidTemplateString.new(string)
+  return setmetatable({ template = LiquidTemplate:parse(string) },
+                      liquid_template_string_mt)
+end
+
+function LiquidTemplateString:render(context)
+  return self.template:render(
+    LiquidInterpreterContext:new(context),
+    liquid_filter_set,
+    liquid_resource_limit
+  )
+end
+
+local PlainTemplateString = {}
+local plain_template_string_mt = { __index = PlainTemplateString }
+
+function PlainTemplateString.new(string)
+  return setmetatable({ string = string }, plain_template_string_mt)
+end
+
+function PlainTemplateString:render()
+  return self.string
+end
+
+local template = {
+  plain = PlainTemplateString,
+  liquid = LiquidTemplateString
+}
+
+local _M = {}
+
+--- Initialize a template
+-- Initialize a liquid or a plain text template according to the given type.
+-- @tparam string value String to construct the template from
+-- @tparam string type Render the template as this type.
+--   Can be 'liquid' or 'plain'
+-- @treturn a template string
+function _M.new(value, type)
+  return template[type].new(value)
```

Comment:
This will crash when it gets unknown type. It should handle missing types and return `nil, err` IMO.

Reply:
Changed :+1: